src/main/java/com/google/api/codegen/util/py/PythonTypeTable.java

```diff
@@ -82,19 +94,84 @@ public String getAndSaveNicknameFor(String fullName) {
 
   @Override
   public String getAndSaveNicknameFor(TypeName typeName) {
-    return dynamicTypeTable.getAndSaveNicknameFor(typeName);
+    return typeName.getAndSaveNicknameIn(this);
   }
 
   @Override
   public String getAndSaveNicknameFor(TypeAlias alias) {
-    return dynamicTypeTable.getAndSaveNicknameFor(alias);
+    if (!alias.needsImport()) {
```

Comment:
Can you copy over the comment giving examples, as well? It's very hard to read this code without examples, IMO.